Makefile

```diff
@@ -434,3 +405,96 @@ site-serve: .$(DOCKER_SITE_BUILDER_IMAGE).image.stamp
 	touch $@
 
 docs: site-build
+
+#
+# CRD et. al code generation
+#
+
+CRD_BASE_DIR := config/crd/bases
+HELM_TOP_DIR := deployment/helm
+
+# Location to install external generation tool binaries
+BUILD_AUX ?= $(shell pwd)/build-aux
+LOCALBIN ?= $(BUILD_AUX)/bin
+$(LOCALBIN):
+	mkdir -p $(LOCALBIN)
+
+## External generation tool binaries
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+CLIENT_GEN ?= $(LOCALBIN)/client-gen
+
+GIT_CLONE    := git clone
+GIT_CHECKOUT := git checkout
+
+# External generation tool versions
+CONTROLLER_TOOLS_VERSION ?= v0.12.0
+CLIENT_GEN_VERSION ?= v0.28.1
+CODE_GENERATOR_REPO ?= https://github.com/kubernetes/code-generator
+CODE_GENERATOR_VERSION ?= v0.28.1
```

Comment:
This branch name is the one that we create when cloning the repo. Should we rather do the cloning based on the upstream `release-1.28` branch?

`$(GIT_CLONE) $(CODE_GENERATOR_REPO) -b $(CODE_GENERATOR_VERSION) origin/release-1.28?`

The reason is, if we are following particular k8s version, perhaps we should stick to that version of the code generators here too to avoid any surprises that may happen on the master/main branch?

Reply:
Based on my own and @marquiz's past experiment with controller-gen's stability, I deliberately want to version-lock `code-generator` to an exact version which is known to work for us. It might be better/more API-stable now and I am willing to give it a try... once we have figured out the exact details of how to switch from using the deprecated `generate-groups.sh` to `kbe_codegen.sh`. I took a look at that initially during this work, but it was very much not a drop-in replacement, so I decided against it as this particular problem is IMO extremely low on our TODO list at the moment.

Reply:
> Based on my own and @marquiz's past experiment with controller-gen's stability, I deliberately want to version-lock `code-generator` to an exact version which is known to work for us.

Isn't it the same with the command I suggested above? It is using [release-1.28](https://github.com/kubernetes/code-generator/tree/release-1.28) release branch of the controller-gen right?

Reply:
I don't think `release-1.28` would be version-/sha1-locked. If/when a `1.28.2` or any other `1.28.x` gets tagged it will point to the tip of `release-1.28`. IOW, `release-1.28` will receive as yet unknown commits.